File: dart/algorithms/searching/binary_search.dart
// binary_search.dart

int binarySearch(List<int> array, int target) {
  int left = 0;
  int right = array.length - 1;

  while (left <= right) {
    int mid = (left + right) ~/ 2;

    if (array[mid] == target) {
      return mid; // Return the index if the target is found
    } else if (array[mid] < target) {
      left = mid + 1;
    } else {
      right = mid - 1;
    }
  }

  return -1; // Return -1 if the target is not found
}

void main() {
  List<int> array = [2, 4, 6, 8, 10, 12, 14];
  int target = 8;

  int result = binarySearch(array, target);

  if (result != -1) {
    print("Element $target found at index $result.");
  } else {
    print("Element $target not found in the array.");
  }
}

File: go/algorithms/searching/binary_search.go
// binary_search.go

package main

import "fmt"

func binarySearch(arr []int, target int) int {
    low, high := 0, len(arr)-1

    for low <= high {
        mid := (low + high) / 2
        midValue := arr[mid]

        if midValue == target {
            return mid
        } else if midValue > target {
            high = mid - 1
        } else {
            low = mid + 1
        }
    }

    return -1
}

func main() {
    sortedArray := []int{1, 2, 3, 4, 5, 6, 7, 8, 9, 10}
    target := 7

    result := binarySearch(sortedArray, target)

    if result != -1 {
        fmt.Printf("Element %d found at index %d\n", target, result)
    } else {
        fmt.Printf("Element %d not found in the array\n", target)
    }
}
